dora/core/common/src/main/java/alluxio/uri/UfsUrl.java

```diff
@@ -0,0 +1,392 @@
+ * <p>
+ * A UFS URL:
+ * <ul>
+ *   <li><b>always</b> has a scheme that is not CONSTANTS SCHEME, e.g. {@code s3a:};</li>
+ *   <li><b>optionally,</b> can have an authority.
+ *   The authority can be made up of a hostname and a port,
+ *   e.g. {@code localhost:80}, but can be an arbitrary string,
+ *   e.g. a bucket name in an object store;
+ *   </li>
+ *   <li><b>always</b> has an absolute path.
+ *   A path is a list of path segments that are separated by {@code /}.
+ *   </li>
+ * </ul>
+ * Examples of valid UFS URL's:
+ * <ul>
+ *   <li>s3a://bucket1/file</li>
+ *   <li>hdfs://namenode:9083/</li>
+ *   <li>file:///home/user/file</li>
+ * </ul>
+ * Examples of invalid UFS URL's:
+ * <ul>
+ *   <li>alluxio://localhost:19998/</li>
+ *   <li>/opt/alluxio</li>
+ * </ul>
+ */
+public class UfsUrl {
+
+  public static final String SCHEME_SEPARATOR = "://";
+  public static final String DOUBLE_SLASH_SEPARATOR = "//";
+  public static final String SLASH_SEPARATOR = "/";
+  public static final String COLON_SEPARATOR = ":";
+
+  private final UfsUrlMessage mProto;
+
+  private static class Parser {
+    private final String mScheme;
+    private final String mAuthority;
+    private final List<String> mPathComponents;
+
+    private Parser(String inputUrl) {
+      Preconditions.checkArgument(inputUrl != null && !inputUrl.isEmpty(),
+          "The input url is null or empty, please input a valid url.");
+      String scheme = null;
+      String authority = null;
+      String path = null;
+
+      int start = 0;
+      int schemeSplitIndex = inputUrl.indexOf(SCHEME_SEPARATOR, start);
+      if (schemeSplitIndex == -1) {
+        scheme = "";
+      } else {
+        scheme = inputUrl.substring(start, schemeSplitIndex);
+        start += scheme.length();
+      }
+      Preconditions.checkArgument(!scheme.equalsIgnoreCase("alluxio"),
+          "Alluxio 3.x no longer supports alluxio:// scheme,"
+              + " please input the UFS path directly like hdfs://host:port/path");
+
+      // unified address "://" and "//"
+      while (start < inputUrl.length() && inputUrl.charAt(start) == COLON_SEPARATOR.charAt(0))  {
+        start++;
+      }
+
+      if (inputUrl.startsWith(DOUBLE_SLASH_SEPARATOR, start)
+          && start + DOUBLE_SLASH_SEPARATOR.length() < inputUrl.length()) {
+        start += DOUBLE_SLASH_SEPARATOR.length();
+        int authoritySplitIndex = inputUrl.indexOf(SLASH_SEPARATOR, start);
+        if (authoritySplitIndex == -1)  {
+          authority = inputUrl.substring(start);
+        } else {
+          authority = inputUrl.substring(start, authoritySplitIndex);
+        }
+      } else {
+        authority = "";
+      }
+
+      start += authority.length();
+      // remove the fronting slash, if any.
+      while (start < inputUrl.length() && inputUrl.charAt(start) == SLASH_SEPARATOR.charAt(0)) {
+        start++;
+      }
+      path = FilenameUtils.normalizeNoEndSeparator(inputUrl.substring(start));
```

Comment:
in a later PR ideally we avoid this copy

Reply:
yes, I forget to leave a comment here, copy twice is terrible.